### PR TITLE
Make the reset password page unavailable if expired.

### PR DESCRIPTION
**Pull request by wsy19961129**
As this issue #112 is only a UI bug, so I simply use base64 to encode the expired time and attach it to the end of the url. Thus the UI would read this and check if the page is expired.

I don't think this will bring security problems as even if a user modify the url and gain access to an expired page, he/she still cannot reset the password as the backend will check that again;

In addition, make a query via Graphql everytime before rendering the UI will bring extra cost, which I think is meaningless.

### CODE REVIEW

File: packages/itmat-interface/src/graphql/resolvers/userResolvers.ts [hunk 1]
```diff
@@ -34,6 +34,40 @@ export const userResolvers = {
...
+            try {
+                email = await decryptEmail(args.encryptedEmail, salt, iv);
+            } catch (e) {
+                throw new ApolloError('Token is not valid.');
+            }
+
+            /* check whether username and token is valid */
+            /* not changing password too in one step (using findOneAndUpdate) because bcrypt is costly */
+            const TIME_NOW = new Date().valueOf();
+            const ONE_HOUR_IN_MILLISEC = 60 /* minutes per hr */ * 60 /* sec per min */ * 1000 /* milli per unit */;
```
Comment (fguitton) [inline packages/itmat-interface/src/graphql/resolvers/userResolvers.ts:55]: This is not consistent with the same line below

File: packages/itmat-interface/src/graphql/resolvers/userResolvers.ts [hunk 2]
```diff
@@ -343,7 +374,8 @@ export const userResolvers = {
             /* check whether username and token is valid */
             /* not changing password too in one step (using findOneAndUpdate) because bcrypt is costly */
             const TIME_NOW = new Date().valueOf();
-            const ONE_HOUR_IN_MILLISEC = 60 /* minutes per hr */ * 60 /* sec per min */ * 1000 /* milli per unit */;
+            // const ONE_HOUR_IN_MILLISEC = 60 /* minutes per hr */ * 60 /* sec per min */ * 1000 /* milli per unit */;
```
Comment (fguitton) [inline packages/itmat-interface/src/graphql/resolvers/userResolvers.ts:384]: Why do we need this comment ? The variable name is already pretty explicit

File: packages/itmat-ui-react/src/components/login/resetPasswordPage.tsx [hunk 3]
```diff
@@ -58,18 +60,19 @@ export const ResetPasswordPage: React.FunctionComponent<ResetPasswordPageProps>
...
                             <br />
                             <div>
                                 <Form onFinish={(variables) => resetPassword({
                                     variables: {
                                         ...variables,
-                                        encryptedEmail,
-                                        token
+                                        encryptedEmail: encryptedEmail,
+                                        token: token,
+                                        queryValidation: false
```
Comment (fguitton) [inline packages/itmat-ui-react/src/components/login/resetPasswordPage.tsx:71]: I don't understand what this variable is for ? It doesn't seem to be used

File: packages/itmat-interface/src/graphql/resolvers/userResolvers.ts [hunk 4]
```diff
@@ -316,7 +347,7 @@ export const userResolvers = {
             await userCore.deleteUser(args.userId);
             return makeGenericReponse(args.userId);
         },
-        resetPassword: async (__unused__parent: Record<string, unknown>, { encryptedEmail, token, newPassword }: { encryptedEmail: string, token: string, newPassword: string }): Promise<IGenericResponse> => {
+        resetPassword: async (__unused__parent: Record<string, unknown>, { encryptedEmail, token, newPassword }: { encryptedEmail: string, token: string, newPassword: string, queryValidation: boolean }): Promise<IGenericResponse> => {
```
Comment (fguitton) [inline packages/itmat-interface/src/graphql/resolvers/userResolvers.ts:357]: This `queryValidation` variable doesn't appear to be used anywhere.

File: packages/itmat-interface/src/graphql/resolvers/userResolvers.ts (hunk 1 above)
Comment (wsy19961129) [reply packages/itmat-interface/src/graphql/resolvers/userResolvers.ts:52]: I simplified the validation process. As I think this part only needs to validate whether the link is expired, thus we do not need to validate other info, like password.

File: packages/itmat-ui-react/src/components/login/resetPasswordPage.tsx [hunk 5]
```diff
@@ -17,6 +18,18 @@ export const ResetPasswordPage: React.FunctionComponent<ResetPasswordPageProps>
...
+    const { loading, error } = useQuery(GQLRequests.VALIDATE_RESET_PASSWORD, { variables: {
+        encryptedEmail: encryptedEmail,
+        token: token
+    }});
+    if (loading) { return <LoadSpinner />; }
+    if (error) {
+        return <div className={css.login_wrapper}>
+            <div className={css.login_box}>
+                <h1>The link is invalid. Please make a new request.</h1>
+            </div>
```
Comment (fguitton) [inline packages/itmat-ui-react/src/components/login/resetPasswordPage.tsx:30]: Could we add a button permitting return to the login page. Something like :
```jsx
<Button onClick={() => {
    history.push('/');
}}>
    Go back to login
</Button>
```

File: packages/itmat-interface/src/graphql/resolvers/userResolvers.ts (hunk 1 above)
Comment (fguitton) [reply packages/itmat-interface/src/graphql/resolvers/userResolvers.ts:52]: I was more thinking about the inline-comments. The name of the variable is expressive enough that we don't need all the `/* ... */` 😄

File: packages/itmat-ui-react/src/components/login/resetPasswordPage.tsx (hunk 5 above)
Comment (wsy19961129) [reply packages/itmat-ui-react/src/components/login/resetPasswordPage.tsx:30]: Sure!

**Pull request merged**